### DISCUSSION TITLE
Asynchronous calls (Promise/Observable) in 'appStateChange' handler

### DISCUSSION THREAD

**Discussion by ghost** [ionic-team/capacitor#4833] [Q&A]
I am working on an Angular 12 / Capacitor 3.1.1 app.

I am using the official App pligin in order to react to app state changes:

``` ts
App.addListener('appStateChange', this.onAppStateChange.bind(this))
```
The handler function loosk something like this:

``` ts
private onAppStateChange(state: AppState): void {
    this._zone.run(() => {
      if (state.isActive) {
        // do stuff and load app state
      } else {
       // do other stuff and save app state
      }
    });
  }
```
Loading and saving app state can potintially include calling asynchronous functions like `Storage.get` or `Storage.set`.
Since `onAppStateChange` has to return `void` in order to work with Capacitor, how am I supposed to work with asynchronous calls and wait for them to finish?

Basically what can happen is that the asynchronous call is made, but the Promise finishes long after the app has been suspended.

How am I supposed to work with `Storage` in this case?
